pyccel/decorators.py

```diff
@@ -139,7 +173,19 @@ class KernelAccessor:
         def __init__(self, f):
             self._f = f
         def __getitem__(self, args):
-            return self._f
+            num_blocks, num_threads = args
+            def internal_loop(*args, **kwargs):
+                """
+                The internal loop for kernel execution.
+
+                The internal loop for kernel execution.
+                """
+                for b in range(num_blocks):
+                    for t in range(num_threads):
+                        self._f.__globals__['cu'] = CudaThreadIndexing(b, t)
```

Comment:
What is the global `cu` ? I don't think this name will work when you try to retrieve the thread dimension and block dimension

Reply:
[ec738b3](https://github.com/pyccel/pyccel-cuda/pull/69/commits/ec738b3b55f3f9ad23b227262f436f132477af43)

Reply:
This change doesn't look right. I just want you to rename `cu` as the current implementation doesn't work. I don't think your new implementation works either.

Please add a test to make sure the threads/blocks/etc can actually be retrieved

Reply:
Do you understand what this line does and what `self._f.__globals__` is?

Reply:
```py
# pylint: disable=missing-function-docstring, missing-module-docstring
from pyccel.decorators import kernel
from pyccel import cuda

@kernel
def say_hello(its_morning : bool):
    if(its_morning):
        print("Hello and Good morning")
    else:
        print("Hello and Good afternoon")

def f():
    its_morning = True
    say_hello[5,5](its_morning)
    cuda.synchronize()

if __name__ == '__main__':
    f()
```
he CUDA model gets overridden by the class CudaThreadIndexing, and we encounter the error:
AttributeError: 'CudaThreadIndexing' object has no attribute 'synchronize'
So using a dynamically assigned value to a dictionary key solves that problem, and the new implementation works just fine.

Reply:
I disagree I am pretty sure that this:
```python
@kernel
def say_hello_thread():
     print(cuda.threadIdx)

if __name__ == '__main__':
    say_hello_thread[5,5]()
```
gives different results on Python vs Cuda.

After this PR in Python it should print:
```
0 0 0 0 0 1 1 1 1 1 2 2 2 2 2 3 3 3 3 3 4 4 4 4 4
```
I am pretty sure that your implementation will print:
```
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```

Reply:
> he CUDA model gets overridden by the class CudaThreadIndexing, and we encounter the error:
AttributeError: 'CudaThreadIndexing' object has no attribute 'synchronize'

In that case the class should not be `CudaThreadIndexing` but rather something like `CudaModuleInstance`

Reply:
i this that better 528099f

Reply:
A little but you should not need to do this:
https://github.com/pyccel/pyccel-cuda/commit/528099f81210532cc16f73fa1d79392b55fd4216#diff-fde82864b093f3b7f981abc6ec5374f7d4d5f4e695ed5d11363ee6eec8b2d2dcR14

synchronise should be inside the new class which is why I suggested renaming it `CudaModuleInstance`. With your current solution `cuda.threadIdx` will give the right result the first time it is called but the wrong result if it is called after the import

Reply:
But cuda.threadIdx should be called only inside a kernel call

Reply:
I'm trying to separate the method of CUDA called from the kernel and not from the kernel.

Reply:
57f977e

Reply:
@EmilyBourne

Reply:
or ```py
 for b in range(num_blocks):
                    for t in range(num_threads):
                        cu = CudaThreadIndexing(b, t)
                        cuda_globals = self._f.__globals__['cuda']
                        setattr(cuda_globals, 'threadIdx', cu.threadIdx)
                        setattr(cuda_globals, 'blockIdx', cu.blockIdx)
                        setattr(cuda_globals, 'blockDim', num_threads)
                        self._f(*args, **kwargs)
                        ```

Reply:
Ah! That sounds like a better idea :+1:
Good thinking :)